File: 01-Languages/Haskell/08-Strings/Strings.hs
import Data.Char (toUpper)

greeting :: String
greeting = "Hello"

greetingAsList :: [Char]
greetingAsList = ['H', 'e', 'l', 'l', 'o']

sameValue :: Bool
sameValue = greeting == greetingAsList

shout :: String -> String
shout = map toUpper

countLetter :: Char -> String -> Int
countLetter c = length . filter (== c)

reversedGreeting :: String
reversedGreeting = reverse greeting

main :: IO ()
main = do
    putStrLn ("greeting == greetingAsList: " ++ show sameValue)
    putStrLn ("shout \"hello\" = " ++ shout "hello")
    putStrLn ("countLetter 'l' \"hello world\" = " ++ show (countLetter 'l' "hello world"))
    putStrLn ("reversedGreeting = " ++ reversedGreeting)
    putStrLn ("length \"hello\" = " ++ show (length "hello"))
    putStrLn ("\"hello\" ++ \" \" ++ \"world\" = " ++ ("hello" ++ " " ++ "world"))
    putStrLn ("words \"the quick brown fox\" = " ++ show (words "the quick brown fox"))
    putStrLn ("unwords [\"the\",\"quick\",\"brown\",\"fox\"] = " ++ unwords ["the", "quick", "brown", "fox"])

File: 01-Languages/Perl/08-Strings/strings.pl
#!/usr/bin/env perl
use strict;
use warnings;

my $name = "World";

# Interpolation only happens in double quotes.
print "double: Hello $name\n";
print 'single: Hello $name', "\n";

# Regex matching with =~ and m//
my $text = "The quick brown fox jumps over 42 lazy dogs";
if ($text =~ /(\d+)/) {
    print "found number: $1\n";
}

my @words = ($text =~ /(\w+)/g);
print "word count: ", scalar(@words), "\n";

# Substitution with s///
(my $censored = $text) =~ s/fox/cat/;
print "original:  $text\n";
print "censored:  $censored\n";

# Global substitution + capture groups
(my $swapped = "2026-07-19") =~ s/(\d+)-(\d+)-(\d+)/$3\/$2\/$1/;
print "date swap: $swapped\n";
